server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java

```diff
@@ -22,15 +22,19 @@
 public interface ThreadContextStatePropagator {
     /**
      * Returns the list of transient headers that needs to be propagated from current context to new thread context.
-     * @param source current context transient headers
+     *
+     * @param source          current context transient headers
+     * @param isSystemContext if the propagation is for system context.
      * @return the list of transient headers that needs to be propagated from current context to new thread context
      */
-    Map<String, Object> transients(Map<String, Object> source);
+    Map<String, Object> transients(Map<String, Object> source, boolean isSystemContext);
 
     /**
      * Returns the list of request headers that needs to be propagated from current context to request.
-     * @param source current context headers
+     *
+     * @param source          current context headers
+     * @param isSystemContext if the propagation is for system context.
      * @return the list of request headers that needs to be propagated from current context to request
      */
-    Map<String, String> headers(Map<String, Object> source);
+    Map<String, String> headers(Map<String, Object> source, boolean isSystemContext);
```

Comment:
true/false can be hard to correctly interpret, I'd prefer enum instead of boolean to improve readability.  What do you think?

```java
propigator.headers(context.headers, false)
propigator.headers(context.headers, true)
// VS
propigator.headers(context.headers, ThreadContextType.System)
propigator.headers(context.headers, ThreadContextType.Default)
```

Reply:
> true/false can be hard to correctly interpret, I'd prefer enum instead of boolean to improve readability. What do you think?

This is very good point, I also thought about that but sided with boolean (to align with ThreadContextStruct  that uses the same semantics). May be we could keep it simple at the moment.